File: webui_utilities/pages/stepper_page.py
import allure
from .base_page import BasePage


class StepperPage(BasePage):
    """Page Object for Stepper/Multi-step form page."""

    # Selectors - updated for Material Stepper
    STEPPER_CONTAINER = "mat-stepper, .mat-stepper"
    STEP_1_HEADER = "button[type='button']:nth-of-type(1)"
    STEP_2_HEADER = "button[type='button']:nth-of-type(2)"
    STEP_3_HEADER = "button[type='button']:nth-of-type(3)"
    ACTIVE_STEP = ".mat-step-header-selected, [aria-selected='true']"
    NEXT_BUTTON = "button:has-text('Next')"
    PREVIOUS_BUTTON = ".mat-stepper-previous"
    FINISH_BUTTON = "button:has-text('Finish'), button:has-text('Submit')"
    STEP_CONTENT = ".mat-step-content"
    STEP_FORM = "form"

    def is_stepper_page_loaded(self) -> bool:
        """Verify stepper page is loaded."""
        with allure.step("Verify stepper page is loaded"):
            return self.is_element_visible(self.STEPPER_CONTAINER)

    def click_step_1(self):
        """Click on step 1 header."""
        with allure.step("Navigate to Step 1"):
            self.click(self.STEP_1_HEADER)

    def click_step_2(self):
        """Click on step 2 header."""
        with allure.step("Navigate to Step 2"):
            self.click(self.STEP_2_HEADER)

    def click_step_3(self):
        """Click on step 3 header."""
        with allure.step("Navigate to Step 3"):
            self.click(self.STEP_3_HEADER)

    def get_active_step(self) -> str:
        """Get currently active step."""
        with allure.step("Get currently active step"):
            try:
                return self.get_text(self.ACTIVE_STEP)
            except:
                return None

    def click_next(self):
        """Click next button."""
        with allure.step("Click next button to advance stepper"):
            self.click(self.NEXT_BUTTON)

    def click_previous(self):
        """Click previous button."""
        with allure.step("Click previous button to go back in stepper"):
            self.click(self.PREVIOUS_BUTTON)

    def click_finish(self):
        """Click finish button."""
        with allure.step("Click finish button to complete stepper"):
            self.click(self.FINISH_BUTTON)

    def is_next_button_visible(self) -> bool:
        """Check if next button is visible."""
        with allure.step("Verify next button is visible"):
            return self.is_element_visible(self.NEXT_BUTTON)

    def is_previous_button_visible(self) -> bool:
        """Check if previous button is visible."""
        with allure.step("Verify previous button is visible"):
            return self.is_element_visible(self.PREVIOUS_BUTTON)

    def is_finish_button_visible(self) -> bool:
        """Check if finish button is visible."""
        with allure.step("Verify finish button is visible"):
            return self.is_element_visible(self.FINISH_BUTTON)

    def navigate_to_step(self, step_number: int):
        """Navigate to specific step (1, 2, or 3)."""
        with allure.step(f"Navigate to step {step_number}"):
            if step_number == 1:
                self.click_step_1()
            elif step_number == 2:
                self.click_step_2()
            elif step_number == 3:
                self.click_step_3()
            self.page.wait_for_load_state("networkidle")

File: src/pages/StepperPage.ts
import { Page } from '@playwright/test';
import { BasePage } from './BasePage';
import { config } from '../utils/config';

/**
 * Page Object for Stepper Page
 */
export class StepperPage extends BasePage {
  // Selectors for stepper elements
  private readonly stepperContainer = '[role="tablist"], .stepper, .mat-stepper';
  private readonly stepLabel = '.mat-step-label, [role="tab"]';
  private readonly nameInput = 'input[aria-label*="name"], input[placeholder*="Name"], #name';
  private readonly addressInput = 'textarea[aria-label*="address"], input[aria-label*="address"], textarea, #address';
  private readonly nextButton = 'button:has-text("Next"), button[aria-label*="Next"]';
  private readonly previousButton = 'button:has-text("Previous"), button:has-text("Back")';
  private readonly doneButton = 'button:has-text("Done"), button[aria-label*="Done"]';
  private readonly finishButton = 'button:has-text("Finish")';
  private readonly stepContent = '.mat-step-content, [role="tabpanel"]';

  constructor(page: Page) {
    super(page);
  }

  /**
   * Navigate to stepper page
   */
  async goto(): Promise<void> {
    await this.navigate(`${config.baseUrl}${config.stepperUrl}`);
  }

  /**
   * Check if stepper page is loaded
   */
  async isStepperPageLoaded(): Promise<boolean> {
    return await this.isElementVisible(this.stepperContainer);
  }

  /**
   * Get number of steps
   */
  async getStepCount(): Promise<number> {
    return await this.page.locator(this.stepLabel).count();
  }

  /**
   * Get current step label
   */
  async getCurrentStepLabel(): Promise<string | null> {
    // Find the active step
    const activeStep = this.page.locator(`${this.stepLabel}[aria-selected="true"], ${this.stepLabel}.active`);
    return await activeStep.textContent();
  }

  /**
   * Fill name on first step
   */
  async fillName(name: string): Promise<void> {
    await this.fill(this.nameInput, name);
  }

  /**
   * Get name field value
   */
  async getNameValue(): Promise<string | null> {
    return await this.page.inputValue(this.nameInput);
  }

  /**
   * Fill address on second step
   */
  async fillAddress(address: string): Promise<void> {
    await this.fill(this.addressInput, address);
  }

  /**
   * Get address field value
   */
  async getAddressValue(): Promise<string | null> {
    try {
      return await this.page.inputValue(this.addressInput);
    } catch {
      return await this.getText(this.addressInput);
    }
  }

  /**
   * Click next button to go to next step
   */
  async clickNext(): Promise<void> {
    await this.click(this.nextButton);
    await this.page.waitForLoadState('domcontentloaded');
  }

  /**
   * Click previous button to go to previous step
   */
  async clickPrevious(): Promise<void> {
    // Try previous first, then back button
    try {
      await this.click(this.previousButton);
    } catch {
      // Button might not exist on first step
    }
    await this.page.waitForLoadState('domcontentloaded');
  }

  /**
   * Click done button (usually on last step)
   */
  async clickDone(): Promise<void> {
    const isDone = await this.isElementVisible(this.doneButton);
    if (isDone) {
      await this.click(this.doneButton);
      await this.page.waitForLoadState('domcontentloaded');
    }
  }

  /**
   * Click finish button
   */
  async clickFinish(): Promise<void> {
    await this.click(this.finishButton);
    await this.page.waitForLoadState('domcontentloaded');
  }

  /**
   * Complete entire stepper flow
   */
  async completeStepperFlow(name: string, address: string): Promise<void> {
    // Step 1: Name
    await this.fillName(name);
    await this.clickNext();

    // Step 2: Address
    await this.fillAddress(address);
    await this.clickNext();

    // Final step: Done
    await this.clickDone();
  }

  /**
   * Check if next button is enabled
   */
  async isNextButtonEnabled(): Promise<boolean> {
    return await this.isElementEnabled(this.nextButton);
  }

  /**
   * Check if previous button is visible
   */
  async isPreviousButtonVisible(): Promise<boolean> {
    return await this.isElementVisible(this.previousButton);
  }

  /**
   * Check if step is completed
   */
  async isStepCompleted(stepIndex: number): Promise<boolean> {
    const step = this.page.locator(`${this.stepLabel}`).nth(stepIndex);
    const ariaCompleted = await step.getAttribute('aria-completed');
    return ariaCompleted === 'true';
  }

  /**
   * Navigate to specific step
   */
  async goToStep(stepIndex: number): Promise<void> {
    const step = this.page.locator(`${this.stepLabel}`).nth(stepIndex);
    await step.click();
    await this.page.waitForLoadState('domcontentloaded');
  }
}
